Restructure modules

ktras opened the pull request:
Closes #64.

bonachea commented on src/caffeine/collective_helpers_m.f90:
IMO this module is nasty. It's not guaranteed by the PRIF spec, and therefore should not be provided/exported by Caffeine to the client / test code as it currently is.

IIUC it's also unnecessary in test code, which appears could trivially be rewritten to stop using it, allowing this to remain a private submodule. Longer term the `prif_co_reduce` implementation could/should be reworked to stop using this ugly hack entirely (which I believe is also a prerequisite to supporting non-blocking `co_reduce`).


Please enter an issue to clean this up, and we can discuss details in our next meeting.

bonachea reviewed (changes requested):
Overall this looks like the wonderful and long-overdue cleanup we discussed.

Added a few questions and minor requests.

ktras commented on src/prif.f90:
```diff
@@ -1,25 +1,655 @@
+  public :: prif_atomic_add, prif_atomic_and, prif_atomic_or, prif_atomic_xor, prif_atomic_cas, prif_atomic_fetch_add
+  public :: prif_atomic_fetch_and, prif_atomic_fetch_or, prif_atomic_fetch_xor, prif_atomic_define, prif_atomic_ref
+
+  type, bind(C) :: handle_data
```
Thanks for the feedback. The change has been made.

ktras commented on src/prif.f90:
```diff
@@ -1,25 +1,655 @@
+    integer(c_intmax_t) :: lcobounds(15), ucobounds(15)
+  end type
+
+  type :: prif_coarray_handle
```
Change made.

ktras commented on src/prif.f90:
```diff
@@ -1,25 +1,655 @@
+    type(handle_data), pointer :: info
+  end type
+
+  type :: prif_team_type
```
Change made.

ktras commented on src/caffeine/prif_private_s.f90:
```diff
@@ -0,0 +1,22 @@
+  type(c_ptr) :: non_symmetric_heap_mspace
+
+  ! TODO: Do these types need to be public? They are not used as a type of any of the args
+  ! for the public procedures
```
Of course, thanks for the reminder as to why they must be public. Change made.

ktras commented on src/prif.f90:
I made no changes to the contents when moving them. Just cut and paste from one module to the other.

ktras commented:
I just moved the definition of `handle_data` to the bottom of the `prif` module. `prif_coarray_handle` and `prif_team_type` type definitions cannot be moved to the same location because there are arguments of those types above.

bonachea commented:
> I just moved the definition of `handle_data` to the bottom of the `prif` module. `prif_coarray_handle` and `prif_team_type` type definitions cannot be moved to the same location because there are arguments of those types above.

I thought we tried this in our prototyping meeting and decided that at least one compiler rejected the declaration of `prif_coarray_handle` when the `handle_data` pointer field was an incomplete type? What compilers have you successfully tested against?

Of course what ultimately matters most is what flang-new will accept, which I'm guessing is currently "none of this". On a related note, I think this PR needs to update the table in the top-level README where we track major modern Fortran features the Caffeine implementation requires. We at least need to add "Submodule support [1]" (which I believe was added in F08?), and probably remove the [1] from `contiguous` attribute, which is now "baked into" the PRIF API. There may be other updates needed.

bonachea reviewed (changes requested):
Thanks for the recent changes, which look good.

Some unresolved tasks remain.

ktras commented:
> I thought we tried this in our prototyping meeting and decided that at least one compiler rejected the declaration of `prif_coarray_handle` when the `handle_data` pointer field was an incomplete type? What compilers have you successfully tested against?

When compiling the whole software stack, the only compiler we are currently developing with is gfortran. 

Based on this comment, I tried compiling just the `prif` module (with the `handle_data` type definition at the bottom of the module) with the following compilers and with the following results:

|   Compiler  |  Result  | 
|------------|---------|
| gfortran      |     compiles        | 
| ifx                |     compiles        | 
| cray ftn       |     compiles        | 
| nvfortran    |     compiles        | 
| nagfor        |  doesn't compile for different reason (doesn't have `c_ptrdiff_t` support). |

ktras commented:
@bonachea Thank you for the further feedback. I have answered your compiler question above and then updated the PR in reference to your requested changes around `collective_helpers_m` and replied to that comment above. I believe I have addressed all of your review comments to this point. Please let me know about any further feedback, thanks!

bonachea reviewed (approved):
LGTM! Thanks for the overhaul and recent fixes.

bonachea commented:
Wooops, I just noticed that doc-generator.md also needs updates as a result of this reorg, assuming we are still maintaining the FORD docs.

bonachea commented on src/caffeine/prif_private_s.f90:
```diff
@@ -0,0 +1,10 @@
+! Copyright (c), The Regents of the University of California
+! Terms of use are as specified in LICENSE.txt
+submodule(prif) prif_private_s
+  implicit none
```
The Fortran spec rules seem potentially ambiguous regarding the default accessibility behavior of a submodule.

Let's add an explicit private here, both to ensure all compilers treat this submodule's data as private to the `prif` implementation and for documentation purposes.

```suggestion
  implicit none
  private
```

ktras replied:
@bonachea Section `8.5.2 Accessibility attribute` of the Fortran 2023 Standard has a constraint that disallows this.
`C817 An access-spec shall appear only in the specification-part of a module.`

everythingfunctional replied:
> The Fortran spec rules seem potentially ambiguous regarding the default accessibility behavior of a submodule.

Only entities declared in a module are accessible via a `use` of that module. All submodule contents are private implementation details.

bonachea replied:
> Only entities declared in a module are accessible via a `use` of that module. All submodule contents are private implementation details.

Thanks for the confirmation Brad. I suspected this was the case, but could not find confirmation on a brief skim of the relevant spec sections.

bonachea reviewed (changes requested):
After additional thought, I've added two further minor requests.

everythingfunctional reviewed (approved):
LGTM, especially since the changes to the tests were so minimal.

bonachea reviewed (approved):
LGTM! Thanks for this valuable cleanup
